zookeeper-server/src/test/java/org/apache/zookeeper/server/backup/RestorationToolTest.java

```diff
@@ -294,4 +310,54 @@ public int compare(File o1, File o2) {
   private boolean getRandomBoolean(float p) {
     return random.nextFloat() < p;
   }
+
+  @Test
+  public void testRestoreToZxidByCommandLine() throws IOException {
+    //Restore to a random zxid
+    int restoreZxid = random.nextInt(txnCnt);
+    RestoreFromBackupTool restoreTool = new RestoreFromBackupTool();
+    CommandLine cl = Mockito.mock(CommandLine.class);
+    when(cl.hasOption(RESTORE_ZXID_OPTION)).thenReturn(true);
+    when(cl.hasOption(BACKUP_STORE_OPTION)).thenReturn(true);
+    when(cl.hasOption(SNAP_DESTINATION_OPTION)).thenReturn(true);
+    when(cl.hasOption(LOG_DESTINATION_OPTION)).thenReturn(true);
+    when(cl.getOptionValue(RESTORE_ZXID_OPTION)).thenReturn(String.valueOf(restoreZxid));
+    when(cl.getOptionValue(BACKUP_STORE_OPTION))
+        .thenReturn("gpfs::" + backupDir.getPath() + ":" + TEST_NAMESPACE);
+    when(cl.getOptionValue(SNAP_DESTINATION_OPTION)).thenReturn(restoreDir.getPath());
+    when(cl.getOptionValue(LOG_DESTINATION_OPTION)).thenReturn(restoreDir.getPath());
+    Assert.assertTrue(restoreTool.runWithRetries(cl));
+    validateRestoreCoverage(restoreZxid);
+
+    //Restore to latest
+    when(cl.getOptionValue(RESTORE_ZXID_OPTION)).thenReturn(BackupUtil.LATEST);
+    Assert.assertTrue(restoreTool.runWithRetries(cl));
+    validateRestoreCoverage(txnCnt);
+  }
+
+  @Test
+  public void testRestoreToTimestampByCommandLine() throws IOException {
+    //Restore to a timestamp in the middle of backup
```

Comment:
`Test restoration using a timestamp recorded in the midpoint of the test ZNode creation`

Also, I think it would be helpful if you could add some JavaDoc above to explain what and why you're doing the following:

```
if (i == txnCnt / 2) {
        timestampInMiddle = System.currentTimeMillis();
      }
```

which leads to me to wonder - is there a reason we're picking the midpoint this way? Are you just trying to record a valid timestamp in the valid backup/timetable range?